File: keys_zhou_lange_2017/spca/compare_spca.jl
using RCall

function compare_spca(proj_type::ASCIIString = "column")

    # set seed
    seed = 2016
    srand(2016)

    tol     = 1e-4    # convergence tolerance
    feastol = 1e-3    # admissible distance to feasibility at convergence 
    quiet   = true    # turn on output? use this for debugging

    # comparison parameters
    R = vec([130 133 126 122 162 178 183 172 138 174 193 165 160 403 173 173 166 217 222 185 148 148 196 156 198])
    k = length(R)     # number of PCs, in simulation we use k = 25

    # need to get breast cancer RNA data from PMA package
    # matrix is annoyingly stored in transpose, so untranspose it
    x = rcopy("library(PMA); data(breastdata); t(breastdata\$rna)")
    n,p = size(x)

    # must center x
    for i = 1:p
        x[:,i] = (x[:,i] - mean(x[:,i])) 
    end

    # now load our SPCA functions
    include("spca.jl")

    # need SVD for warm start
    u,s,v = svd(x)

    # total variance in dataset?
    totvar = trace(x'*x)

    # track previous variance of k-1 components
    kvar0 = 0.0

    # name the "output" variable in order to return SPCA results
    output = 0

    # precompile SPCA by running once and discarding results
    V = v[:,1:1]
    output = spca(x, 1, R[1], proj_type, U=V, quiet=quiet, feastol=feastol, tol=tol) 

    # spca() uses accelerated PD algo with ortho domain constraints
    println("PCs\tNnz\tObj\tdortho\tTime\tVE\tAVE\tPVE\tIter")
    for i = 1:k 

        # V is warm start for SPCA variable U
        V = v[:,1:i]

        # compute SPCA
        tic()
        output = spca(x, i, R[1:i], proj_type, U=V, quiet=quiet, feastol=feastol, tol=tol) 
        mm_time = toq()

        # get sparse loadings
        u = copy(output["U"])

        # xk is matrix of k PCs
        xk = x * u * ((u'*u) \ u')

        # compute various variances
        kvar    = trace(xk'*xk)
        adjvar  = kvar - kvar0
        dortho  = vecnorm(u'*u - I)
        normvar = trace(u'*x'*x*u)

        # print output
        if proj_type == "matrix"
            @printf("%d\t%d\t%3.0f\t%3.3f\t%3.3f\t%3.0f\t%3.0f\t%3.3f\t%d\n", i, countnz(u), normvar, dortho, mm_time, kvar, adjvar, kvar / totvar, output["iter"])
        else
            @printf("%d\t%d\t%3.0f\t%3.3f\t%3.3f\t%3.0f\t%3.0f\t%3.3f\t%d\n", i, countnz(u[:,i]), normvar, dortho, mm_time, kvar, adjvar, kvar / totvar, output["iter"])
        end

        # reset V, kvar0
        fill!(V, 0.0)
        kvar0 = kvar
    end 

    return nothing
end

println("# Results from matrix projection")
compare_spca("matrix")

println("# Results from columnwise projection")
compare_spca()

println("# Remember: added compute cost of SVD warm start in Julia is ~2 sec!")

File: keys_zhou_lange_2019/spca/compare_spca.R
# load libraries 
library(PMA)
library(Matrix)
library(compiler)

# precompile a function to round a number to a correct number of digits
prc = cmpfun(function(x, k) format(round(x, k), nsmall=k))

compare.spca = cmpfun(function(){

    # get dataset
    data(breastdata)

    # RNA data must be transposed before running
    X = t(breastdata$rna)
    p = dim(X)[2]
    x.means = apply(X, 2, mean)
    x = X
    for (i in 1:p){ x[,i] = x[,i] - x.means[i] }

    # set algo parameters
    sumabsv = 8
    niter   = 1000
    trace   = F
    orth    = T
    center  = F 
    vpos    = F
    vneg    = F
    cmp.pve = F
    K       = 25 
    v       = NULL
    seed    = 2016
    totvar  = sum(diag(t(x) %*% x))
    kvar0   = 0

    # header for output
    cat("#PC\tNnz\tObj\td(ortho)\tTime\tPVE\tAPVE\tCPVE\n")

    # set seed and then run algo once for each model size
    # ensure warm start each time
    set.seed(seed)
    for ( i in 1:K ){

        # time SPC 
        start.time = proc.time()
        my.spc = SPC(x, sumabsv=sumabsv, niter=niter, trace=trace, orth=orth, center=center, vpos=vpos, vneg=vneg, K=i, v=v, compute.pve = cmp.pve)
        stop.time = proc.time() - start.time

        # v is matrix of sparse loading vectors 
        v = as.matrix(my.spc$v)

        # xk is matrix of PCs 
        xk = x %*% v %*% solve( t(v) %*% v, t(v))

        # compute various variances
        kvar = sum(diag(t(xk) %*% xk))
        adjvar = kvar - kvar0
        dortho = sqrt(sum(t(v) %*% v - diag(i))^2)
        normvar = sum(diag(t(v) %*% t(x) %*% x %*% v))

        # print output
        cat(paste(i, "\t", nnzero(v[,i]), "\t", round(normvar), "\t", prc(dortho, 3), "\t", prc(stop.time[3], 3), "\t", round(kvar), round(adjvar), prc(kvar / totvar,2), "\n"))

        # save previous variance of k PCs
        kvar0 = kvar
    }
#    return(my.spc)
     return()
})

#my.spc = compare.spca()
compare.spca()
